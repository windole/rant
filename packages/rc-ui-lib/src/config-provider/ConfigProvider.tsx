import React, { CSSProperties, useMemo } from 'react';
import ConfigProviderContext, { INITIAL_STATE } from './ConfigProviderContext';
import { ConfigProviderProps } from './PropsType';

function kebabCase(str: string): string {
  return str
    .replace(/([A-Z])/g, '-$1')
    .toLowerCase()
    .replace(/^-/, '');
}

function mapThemeVarsToCSSVars(themeVars: Record<string, string | number>) {
  const cssVars: Record<string, string | number> = {};
  Object.keys(themeVars).forEach((key) => {
    cssVars[`--rc-${kebabCase(key)}`] = themeVars[key];
  });
  return cssVars;
}

const ConfigProvider: React.FC<ConfigProviderProps> = ({
  className,
  style,
  themeVars,
  tag,
  children,
  ...props
}) => {
  const varStyle = useMemo<CSSProperties | undefined>(() => {
    if (themeVars) {
      return { ...style, ...mapThemeVarsToCSSVars(themeVars) };
    }
    return style;
  }, [themeVars]);
  return (
    <ConfigProviderContext.Provider value={{ ...INITIAL_STATE, ...props }}>
      {React.createElement(
        tag,
        {
          className,
          style: varStyle,
        },
        children,
      )}
    </ConfigProviderContext.Provider>
  );
};

ConfigProvider.defaultProps = {
  tag: 'div',
};

export default ConfigProvider;
