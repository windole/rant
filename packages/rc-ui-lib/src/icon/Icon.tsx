import React, { useContext, useMemo } from 'react';
import classnames from 'classnames';
import { addUnit } from '../utils';
import { IconProps } from './PropsType';
import Badge from '../badge';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';

function isImage(name?: string): boolean {
  return name ? name.indexOf('/') !== -1 : false;
}

const Icon: React.FC<IconProps> = (props) => {
  const { iconPrefix } = useContext(ConfigProviderContext);

  const { dot, tag, name, className, onClick, onTouchStart } = props;

  const imageIcon = isImage(name);

  const classPrefix = useMemo(
    () => props.classPrefix || iconPrefix,
    [props.classPrefix, iconPrefix],
  );

  return (
    <Badge
      dot={dot}
      tag={tag}
      className={classnames(className, classPrefix, imageIcon ? '' : `${classPrefix}-${name}`)}
      style={{
        color: props.color,
        fontSize: addUnit(props.size),
        ...props.style,
      }}
      onClick={onClick}
      onTouchStart={onTouchStart}
      {...props.badge}
    >
      {props?.children}
      {imageIcon && <img className={classnames('rc-icon__image')} src={name} alt={name} />}
    </Badge>
  );
};

Icon.defaultProps = {
  tag: 'i',
};

export default Icon;
