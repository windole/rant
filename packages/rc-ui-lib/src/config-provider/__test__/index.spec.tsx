import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Icon from '../../icon';
import ConfigProvider from '..';

describe('ConfigProvider', () => {
  let wrapper;
  afterEach(() => {
    wrapper.unmount();
  });

  it('when using themeVars prop', () => {
    const themeVars = {
      color: 'red',
    };
    wrapper = mount(
      <ConfigProvider themeVars={themeVars}>
        <Icon name="success" />
      </ConfigProvider>,
    );
    expect(toJson(wrapper.html())).toMatchSnapshot();
  });

  it('when using iconPrefix prop', () => {
    wrapper = mount(
      <ConfigProvider iconPrefix="my-icon">
        <Icon name="success" />
      </ConfigProvider>,
    );
    expect(wrapper.find('i.my-icon-success').exists()).toBeTruthy();
  });
});
