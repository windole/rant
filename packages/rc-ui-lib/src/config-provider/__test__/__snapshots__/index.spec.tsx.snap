// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider when using themeVars prop 1`] = `undefined`;
